### PR TITLE
DATAUP-549 Add xSV parser

### PR DESCRIPTION
**Pull request by MrCreosote**
Excel parser still to do

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -9,4 +9,8 @@ pytest-cov==2.5.1
 hypothesis==6.14.5
 globus-sdk==1.6.1
 python-dotenv==0.14.0
-frozendict==2.0.3
\ No newline at end of file
+frozendict==2.0.3
```
Comment (bio-boris) [inline requirements.txt:12]: Should we use a later version of frozendict or any of these others?

Comment (MrCreosote) [reply requirements.txt:12]: The frozendict version is less than a month old, and the others should be up to date as of a couple days ago, so I think we're ok

File: staging_service/import_specifications/individual_parsers.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            ErrorType.PARSE_FAIL,
+            f"Schema version {version} is larger than maximum processable "
+            + f"version {maximum_version}",
+            spec_source
+        ))
+    return match[1]
+
+
+def _get_datatype(input_: TextIO, spec_source: SpecificationSource, maximum_version: int) -> str:
+    # return is data type
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:59]: What does this comment mean?

File: staging_service/import_specifications/individual_parsers.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+def _strip(putative_str: PRIMITIVE_TYPE) -> PRIMITIVE_TYPE:
+    return putative_str.strip() if isinstance(putative_str, str) else putative_str
+
+
+def _check_for_duplicate_headers(headers: pandas.Index, spec_source: SpecificationSource):
+    seen = set()
+    for name in headers.get_level_values(0):
+        if name in seen:
+            raise _ParseException(Error(
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:78]: It might be more user friendly to collect all duplicate columns instead of just one, but this is fine.

File: staging_service/import_specifications/individual_parsers.py (hunk 2 above)
Comment (MrCreosote) [reply staging_service/import_specifications/individual_parsers.py]: It's cruft from when the return type was more complicated. Removed

File: staging_service/import_specifications/individual_parsers.py (hunk 3 above)
Comment (MrCreosote) [reply staging_service/import_specifications/individual_parsers.py:78]: yeah, I think there's another place where I note we could collect errors, but this should only happen if a user changes the template, so it's probably not worth the effort / complexity

File: staging_service/import_specifications/individual_parsers.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+_HEADER_SEP = ";"
+_EXPECTED_HEADER = f"{_DATA_TYPE} <data_type>{_HEADER_SEP} {_VERSION_STR} <version>"
+_HEADER_REGEX = re.compile(f"{_DATA_TYPE} (\\w+){_HEADER_SEP} {_VERSION_STR} (\\d+)")
+
+
+class _ParseException(Exception):
+    pass
+
+
+def _parse_header(line: str, spec_source: SpecificationSource, maximum_version: int) -> str:
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:38]: 👍

File: staging_service/import_specifications/individual_parsers.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        raise _ParseException(Error(
+            ErrorType.PARSE_FAIL,
+            f"Schema version {version} is larger than maximum processable "
+            + f"version {maximum_version}",
+            spec_source
+        ))
+    return match[1]
+
+
+def _get_datatype(input_: TextIO, spec_source: SpecificationSource, maximum_version: int) -> str:
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:58]: Is this an input stream?

File: staging_service/import_specifications/individual_parsers.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    try:
+        return _parse_header(next(input_).strip(), spec_source, maximum_version)
+    except StopIteration:
+        raise _ParseException(Error(
+            ErrorType.PARSE_FAIL, "Missing data type / version header", spec_source))
+
+
+def _error(error: Error) -> ParseResults:
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:66]: 👍

File: staging_service/import_specifications/individual_parsers.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    return ParseResults(errors = tuple([error]))
+
+
+def _strip(putative_str: PRIMITIVE_TYPE) -> PRIMITIVE_TYPE:
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:70]: 👍

File: staging_service/import_specifications/individual_parsers.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    return putative_str.strip() if isinstance(putative_str, str) else putative_str
+
+
+def _check_for_duplicate_headers(headers: pandas.Index, spec_source: SpecificationSource):
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:74]: 👍

File: staging_service/import_specifications/individual_parsers.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        if name in seen:
+            raise _ParseException(Error(
+                ErrorType.PARSE_FAIL,
+                f"Duplicate column name: {name.split('.')[0]}",
+                spec_source
+            ))
+        seen.add(name)
+
+
+def _check_xsv_row_count(path: Path, expected_count: int, sep: str):
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:86]: `_check_xsv_row_count_as_expected` might be more clear

Comment (bio-boris) [reply staging_service/import_specifications/individual_parsers.py:86]: 👍

File: staging_service/import_specifications/individual_parsers.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                # lets wait and see if that's really needed
+                raise _ParseException(Error(
+                    ErrorType.PARSE_FAIL,
+                    f"Incorrect number of items in line {i + 2}, "
+                    + f"expected {expected_count}, got {count}",
+                    SpecificationSource(path)
+                ))
+
+def _process_dataframe(df: pandas.DataFrame, spec_source: SpecificationSource) -> ParseResult:
+    results = []
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:104]: It's kinda hard to tell what's going on in here 
So it's passing in a dataframe, and converts it to a dictionary but wait, actually it's not a dictionary, its a list of dictionaries. ( I had to test it here https://www.learnpython.org/en/Pandas_Basics )
We could make this more obvious  by doing

`records = df.to_dict(orient='records') # type: List`

and then
for record in records:
    for key in record:
    This part doesn't make sense. Isn't k[0] an index of a string?

I tried this code in https://www.learnpython.org/en/Pandas_Basics and I got a list of dicts with the key being a single character (Unless i copied and pasted this wrong)

File: staging_service/import_specifications/individual_parsers.py (hunk 5 above)
Comment (MrCreosote) [reply staging_service/import_specifications/individual_parsers.py:58]: yep, it's what you get when you `open()` in text mode

File: staging_service/import_specifications/individual_parsers.py (hunk 9 above)
Comment (MrCreosote) [reply staging_service/import_specifications/individual_parsers.py:86]: how about `_validate_xsv_row_count`? Means the same thing, I think, but shorter

File: staging_service/import_specifications/individual_parsers.py (hunk 10 above)
Comment (MrCreosote) [reply staging_service/import_specifications/individual_parsers.py:104]: > So it's passing in a dataframe, and converts it to a dictionary but wait, actually it's not a dictionary, its a list of dictionaries.

Yeah, pandas packed a bit too much functionality into that method IMO.

I've switched things around and added comments, see what you think

Comment (MrCreosote) [reply staging_service/import_specifications/individual_parsers.py:104]: The type hint for that method must be horrendous, it returns like 5 different type structures

File: staging_service/import_specifications/individual_parsers.py (hunk 5 above)
Comment (MrCreosote) [reply staging_service/import_specifications/individual_parsers.py:58]: or `StringIO` or whatever

File: staging_service/import_specifications/individual_parsers.py (hunk 10 above)
Comment (bio-boris) [reply staging_service/import_specifications/individual_parsers.py:104]: Much better, thanks!

File: tests/import_specifications/test_individual_parsers.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    Error,
+    SpecificationSource,
+    ParseResult,
+    ParseResults
+)
+
+from tests.test_app import FileUtil
+
+@fixture(scope="module")
+def temp_dir() -> Generator[Path, None, None]:
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:23]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    with FileUtil() as fu:
+        childdir = Path(fu.make_dir(str(uuid.uuid4()))).resolve()
+
+        yield childdir
+
+    # FileUtil will auto delete after exiting
+
+
+def test_xsv_parse_success(temp_dir: Path):
+    _xsv_parse_success(temp_dir, ',', parse_csv)
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:33]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+    _xsv_parse_success(temp_dir, '\t', parse_tsv)
+
+
+def _xsv_parse_success(temp_dir: Path, sep: str, parser: Callable[[Path], ParseResults]):
+    s = sep
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:38]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+            f"Spec 1{s} Spec 2{s} Spec 3{s} Spec 4\n",
+            f"val1 {s}   val2   {s}    7     {s} 3.2\n",   # test trimming
+            f"val3 {s} val4{s} 1{s} 8.9\n",
+            f"val5 {s}{s}{s} 42.42\n",                     # test missing values w/o whitespace
+            f"val6 {s}      {s}      {s} 3.14\n"           # test missing values w/ whitespace
+        ])
+
+    res = parser(input_)
+
+    assert res == ParseResults(frozendict(
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:53]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+                frozendict({"spec1": "val3", "spec2": "val4", "spec3": 1, "spec4": 8.9}),
+                frozendict({"spec1": "val5", "spec2": None, "spec3": None, "spec4": 42.42}),
+                frozendict({"spec1": "val6", "spec2": None, "spec3": None, "spec4": 3.14}),
+            ])
+        )}
+    ))
+
+
+def test_xsv_parse_success_with_mixed_column(temp_dir: Path):
+    """
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:66]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+    are treated as strings. This is a built in pandas behavior so we have to live with it,
+    unfortunately - at least without making the templates more heavyweight, which we
+    decided not to do for now.
+    """
+    _xsv_parse_success_with_mixed_column(temp_dir, ',', parse_csv)
+    _xsv_parse_success_with_mixed_column(temp_dir, '\t', parse_tsv)
+
+
+def _xsv_parse_success_with_mixed_column(
+    temp_dir: Path, sep: str, parser: Callable[[Path], ParseResults]
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:78]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+            f"Spec 1{s} Spec 2{s} Spec 3{s} Spec 4\n",
+            f"val1 {s} val2{s}    7     {s} 3.2\n",
+            f"val3 {s} val4{s} 1{s} 8.9\n",
+            f"val5 {s} val6{s} int{s} float\n",
+        ])
+
+    res = parser(input_)
+
+    assert res == ParseResults(frozendict(
+        {"other_type": ParseResult(SpecificationSource(input_),
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:95]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+            tuple([
+                frozendict({"spec1": "val1", "spec2": "val2", "spec3": "7", "spec4": "3.2"}),
+                frozendict({"spec1": "val3", "spec2": "val4", "spec3": "1", "spec4": "8.9"}),
+                frozendict({"spec1": "val5", "spec2": "val6", "spec3": "int", "spec4": "float"}),
+            ])
+        )}
+    ))
+
+def test_xsv_parse_fail_no_file(temp_dir: Path):
+    input_ = temp_dir / str(uuid.uuid4())
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:105]: What is going on here with this slash? A path divided by a string?

Comment (bio-boris) [reply tests/import_specifications/test_individual_parsers.py:105]: Oh oops this is `from pathlib import Path` not staging_service.utils.Path

File: tests/import_specifications/test_individual_parsers.py [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+    res = parse_csv(input_)
+
+    assert res == ParseResults(errors=tuple([
+        Error(ErrorType.FILE_NOT_FOUND, source_1=SpecificationSource(input_))
+    ]))
+
+
+def _xsv_parse_fail(
+    temp_dir: Path, lines: list[str], parser: Callable[[Path], ParseResults], message: str
+):
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:124]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+
+    res = parser(input_)
+    expected = ParseResults(errors=tuple([Error(
+        ErrorType.PARSE_FAIL, message, SpecificationSource(input_)
+    )]))
+    assert res == expected
+
+
+def test_xsv_parse_fail_empty_file(temp_dir: Path):
+    _xsv_parse_fail(temp_dir, [], parse_csv, "Missing data type / version header")
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:137]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+
+
+def test_xsv_parse_fail_bad_datatype_header(temp_dir: Path):
+    err = ('Invalid header; got "This is the wrong header", expected "Data type: '
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:141]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+        + '<data_type>; Version: <version>"')
+    _xsv_parse_fail(temp_dir, ["This is the wrong header"], parse_csv, err)
+
+
+def test_xsv_parse_fail_bad_version(temp_dir: Path):
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:146]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+    err = "Schema version 87 is larger than maximum processable version 1"
+    _xsv_parse_fail(temp_dir, ["Data type: foo; Version: 87"], parse_csv, err)
+
+def test_xsv_parse_fail_missing_column_headers(temp_dir: Path):
+    err = "Expected 2 column header rows"
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:151]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+    _xsv_parse_fail(temp_dir, ["Data type: foo; Version: 1\n"], parse_csv, err)
+
+    lines = ["Data type: foo; Version: 1\n", "head1, head2\n"]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+
+def test_xsv_parse_fail_missing_data(temp_dir: Path):
+    err = "No data in file"
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:159]: I wonder if it would be better to say something about headers vs content, because there is data in the file, it just happens that its header data rather than body data.

File: tests/import_specifications/test_individual_parsers.py [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+    lines = [
+        "Data type: foo; Version: 1\n"
+        "head1, head2, head3\n",
+        "Head 1, Head 2, Head 3\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+
+def test_xsv_parse_fail_unequal_rows(temp_dir: Path):
+    err = "Header rows have unequal column counts"
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:169]: 👍

File: tests/import_specifications/test_individual_parsers.py [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+        "Head 1, Head 2, Head 3\n",
+        "1, 2, 3\n",
+        "\n",
+        "1, 2, 3\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+
+def test_xsv_parse_fail_duplicate_headers(temp_dir: Path):
+    err = "Duplicate column name: head3"
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:221]: 👍

Comment (bio-boris) [reply tests/import_specifications/test_individual_parsers.py:221]: Do we need to test with duplicates on both lines?

**Review by bio-boris — APPROVED** [review]
LGTM

File: tests/import_specifications/test_individual_parsers.py (hunk 26 above)
Comment (MrCreosote) [reply tests/import_specifications/test_individual_parsers.py:221]: I do right? it's the second assert for this method. If that's what not you mean can you give an example?

File: tests/import_specifications/test_individual_parsers.py (hunk 24 above)
Comment (MrCreosote) [reply tests/import_specifications/test_individual_parsers.py:159]: How would you phrase it?

File: tests/import_specifications/test_individual_parsers.py (hunk 26 above)
Comment (bio-boris) [reply tests/import_specifications/test_individual_parsers.py:221]: Yes you do, didn't delete that comment before the end of the review.

File: tests/import_specifications/test_individual_parsers.py (hunk 24 above)
Comment (bio-boris) [reply tests/import_specifications/test_individual_parsers.py:159]: It's probably a rare error, but it this error might occur when a user downloads the template and then uploads a blank template instead of filling it out.

Some ideas could be

Import Spec Error:
Template body missing
Template body blank
Only import headers provided, import body missing
No data in file: (Headers found, body missing)
No data in file: Please fill out the import specification template for <app>

Comment (MrCreosote) [reply tests/import_specifications/test_individual_parsers.py]: Updated

**Review by bio-boris — APPROVED** [review]
LGTM
